### ISSUE TITLE
MPI_Errhandler_set_f test fails (related to MPI_VERSION)

### ISSUE THREAD

**Issue by wzamazon** [open-mpi/ompi#11809]
Thank you for taking the time to submit an issue!

### What version of Open MPI are you using? (e.g., v3.0.5, v4.0.2, git branch name and hash, etc.)

main branch

### Describe how Open MPI was installed (e.g., from a source/distribution tarball, from a git clone, from an operating system distribution package, etc.)

compile from source

### If you are building/installing from a git clone, please copy-n-paste the output from `git submodule status`.

N/A

## Details of the problem

I am looking into failures of mtt's intel test suites.

One of the failed tests is `MPI_Errhanler_f`, which emits the following error message:

```
 Errhandler 1 called w/ incorrect communicator
```

I did some debug and found the error is related to `MPI_VERSION`.

In the source of `MPI_Errhandler_f`. there are code like the following in the error handling function:

```
        IF (MPI_VERSION .GE. 4) THEN
            COMMERR = MPI_COMM_SELF
        ELSE
            COMMERR = MPI_COMM_WORLD
        END IF

        IF (COMMIN .NE. COMMERR) THEN
          INFOBUF='Errhandler 1 called w/ incorrect communicator'
          CALL MPITEST_MESSAGE(MPITEST_NONFATAL, INFOBUF)
          FAIL = FAIL + 1
        END IF

```

Basically, if the MPI_VERSION is >= 4, the handle expect the comm to be MPI_COMM_SELF, otherwise it expect the comm to be MPI_COMM_WORLD.

It appears that even though MPI_VERSION in `mpif-constants.h` is 3, the `comm` the err handler got was `MPI_COMM_SELF`.

If I change `MPI_VERSION` to 4 in `mpif-constants.h`, the test will pass. But I am not sure this is the right fix.

I wonder what is the proper value for `MPI_VERSION`?

**Comment by jsquyres** [comment]
We have not updated `MPI_VERSION` to 4 yet because Open MPI does not support all of MPI-4.

Looks like this test will need some other kind of check to determine which communicator should be used.

**Comment by wzamazon** [comment]
> We have not updated MPI_VERSION to 4 yet because Open MPI does not support all of MPI-4.

It looks like part of the OMPI code thinks it should behavior like MPI 4. specifically, there is a MCA parameter: ompi_mpi_compat_mpi3 which is currently default at `false`, which means to behavior like MPI-4 by default.

I think the default value is wrong, and opened https://github.com/open-mpi/ompi/pull/11816 to fix it.

**Comment by bosilca** [comment]
We are striving to move to MPI 4.0, I think backward compatibility shall be explicitly demanded.

**Comment by wzamazon** [comment]
>We are striving to move to MPI 4.0, I think backward compatibility shall be explicitly demanded.

I am not sure I understand. The main branch's `MPI_VERSION` is 3, but part of the code behavior like MPI-4.

**Comment by jsquyres** [comment]
@bosilca He's got a point: `MPI_VERSION==3`, but `ompi_mpi_compat_mpi3` defaults to False.  That feels like a contradiction.

**Comment by bosilca** [comment]
OK, I think the example above is the intended way to use the `MPI_VERSION`. It is unfortunate that there is no generic/simple way to discover MPI capabilities at configure time without specializing the tests to each MPI implementation, but that's a different topic. I withdraw my concern about this change.
